### PR TITLE
Check sender_key matches on inbound encrypted events.

### PR DESCRIPTION
**Pull request by erikjohnston**
If it doesn't then the device lists are probably out of sync so we should resync.

Also add info logging when we receive device updates to help debug missing device lists. Hopefully this shouldn't be tooooo spammy.

### CODE REVIEW

**Comment by richvdh** [comment]
There's a lot of feedback coming back to the effect that our logs are far too spammy, and I have some sympathy with this position. Can you just enable DEBUG for this logger on matrix.org/jki.re/etc?

File: synapse/handlers/federation.py [hunk 1]
```diff
@@ -754,27 +754,61 @@ def __init__(self, hs):
...
+                    if event.content.get("algorithm") == "m.megolm.v1.aes-sha2":
+                        key_name = "curve25519:%s" % (device_id,)
+                        keys = device.get("keys", {}).get("keys", {})
+                        current_key = keys.get(key_name)
+                    else:
+                        current_key = "<unknown algorithm>"
+                else:
+                    current_key = "<no device_id>"
+
+                # We check that one of the keys matches.
```
Comment (richvdh) [inline synapse/handlers/federation.py]: can you enlarge on this? one of which keys? matches what?

File: synapse/handlers/federation.py [hunk 2]
```diff
@@ -754,27 +754,61 @@ def __init__(self, hs):
...
+                    for device in cached_devices
+                    for key in device.get("keys", {}).get("keys", {}).values()
+                ):
+                    logger.info(
+                        "Received event from remote device with different sender key: %s %s %s (current: %s)",
+                        event.sender,
+                        device_id,
+                        sender_key,
+                        current_key,
                     )
```
Comment (richvdh) [inline synapse/handlers/federation.py]: this looks like it's going to be confusing in the situation where we recognise the `sender_key`, but it belongs to a device other than the given `device_id`.

**Comment by erikjohnston** [comment]
> There's a lot of feedback coming back to the effect that our logs are far too spammy, and I have some sympathy with this position. Can you just enable DEBUG for this logger on matrix.org/jki.re/etc?

Me too, but the problem is by the time we realise that we need to get DEBUG logs its a bit late, given that usually the failure happens one and can't be reproduced.

Adding a single log line for each inbound thing that happens to say what we got feels like the right balance for INFO logs to me. If we actually want to be a bit less insane with logging we should look at dropping things like `Received request` and `Sending request` to DEBUG. For context this logging added 392 lines in the past 11 hours (0.04%) on jki.re, while `Received request` added 147480 lines (16%)

**Comment by erikjohnston** [comment]
Hopefully that's a bit clearer

**Pull request merged**